### PR TITLE
feat(gateway-api): Add initial support for tls and grpc routes

### LINKED ISSUE #237: Support other Gateway API route types
Currently k8s_gateway plugin only supports HTTPRoute.

It will most probably be beneficial to others to also support:

- TLSRoute
- GRPCRoute

Ref spec: https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: example
  namespace: default
spec:
  parentRefs:
    - name: cilium-internal
      namespace: infrastructure
  hostnames:
    - "example.test.com"
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    backendRefs:
      - name: example
        port: 8080
```

### PR DESCRIPTION
**Pull request by larivierec**
- use experimental install crds, for tests
- add tests for tls and grpc

would close #237

### CODE REVIEW

**Comment by larivierec** [comment]
hi @networkop , i'm going to have to find a better way to test from what i understand, i'll need a SSL cert in order to do proper integration tests, this might be complicated to do or perhaps i should spoof them?

do you have any ideas?

side note: i fixed the linting and added tls/grpc to the tests controller i was sure i had previously done that but apparently not.

**Comment by networkop** [comment]
> do you have any ideas?

I think doing tests with self-signed certs should be good enough.

**Comment by larivierec** [comment]
just letting you know, i'm having issues running kind on my arm Mac.
i tried using colima, base lima and kind to spawn a cluster and so far nothing has worked.

i was able to generate the cert locally and create with `kubectl`

just letting you know i didn't forget, trying to configure my local env on m2 properly.

**Comment by larivierec** [comment]
You can cancel runs, for now I’ll let you know when we can test the PR

**Comment by larivierec** [comment]
hi @networkop 

I was finally able to get some very basic integration tests done

Below is the output for k8s-gateway plugin accepting the new tls/grpc routes
```
(⎈|local:kube-system)➜  ~ k logs excoredns-k8s-gateway-67bb559895-tt52g
[INFO] plugin/k8s_gateway: Building k8s_gateway controller
[INFO] plugin/k8s_gateway: Starting k8s_gateway controller
[INFO] plugin/k8s_gateway: Waiting for controllers to sync
.:1053
[INFO] plugin/reload: Running configuration SHA512 = efa79d1a2c701696de610932242797b9aa250ece9ac2abd7f81f4f6a94bb12cc91e15a43257b1f0683c883af76301567b914ea403db46feebb6b5734f99d5dde
CoreDNS-1.11.1+k8s_gateway-0.3.4
linux/arm64, go1.21.0,
[DEBUG] plugin/k8s_gateway: Request 2426908976122741341.6026106597144163097. has not matched any zones [foo.org.]
[DEBUG] plugin/k8s_gateway: Adding index myservicea.foo.org for ingress ingress-myservicea
[INFO] 127.0.0.1:54209 - 18519 "HINFO IN 2426908976122741341.6026106597144163097. udp 57 false 512" NOERROR qr,rd,ra 57 0.000683459s
[DEBUG] plugin/k8s_gateway: Adding index myservicetls for tlsRoute myservicetls.gw.foo.org
[DEBUG] plugin/k8s_gateway: Adding index myservicegrpc for grpcRoute myservicegrpc.gw.foo.org
[DEBUG] plugin/k8s_gateway: Adding index myservicea for httpRoute myservicea.gw.foo.org
[DEBUG] plugin/k8s_gateway: Adding index myserviced for httpRoute myserviced.gw.foo.org
[DEBUG] plugin/k8s_gateway: Adding index myserviceb for httpRoute myserviceb.gw.foo.org
[DEBUG] plugin/k8s_gateway: Adding index myservicec for httpRoute myservicec.gw.foo.org
[INFO] plugin/k8s_gateway: Invalid FQDN length: abcd0123456789012345678901234567890123456789012345678901234567890
[DEBUG] plugin/k8s_gateway: Adding index annotation-bad.default for service annotation-bad
[INFO] plugin/k8s_gateway: RFC 1123 conformance failed for FQDN: foo_bar
[DEBUG] plugin/k8s_gateway: Adding index annotation-bad-2.default for service annotation-bad-2
[DEBUG] plugin/k8s_gateway: Adding index good.ok for service annotation-good
[DEBUG] plugin/k8s_gateway: Adding index gateway.default for service gateway
[DEBUG] plugin/k8s_gateway: Adding index gateway-one-istio.default for service gateway-one-istio
[DEBUG] plugin/k8s_gateway: Adding index gateway-two-istio.default for service gateway-two-istio
[DEBUG] plugin/k8s_gateway: Adding index ingress-nginx-controller.default for service ingress-nginx-controller
[DEBUG] plugin/k8s_gateway: Adding index nginxinc-nginx-ingress-controller.default for service nginxinc-nginx-ingress-controller
[DEBUG] plugin/k8s_gateway: Adding index test.default for service test
[DEBUG] plugin/k8s_gateway: Adding index cilium-ingress.kube-system for service cilium-ingress
[INFO] plugin/k8s_gateway: Synced all required resources
```

I had to heavily modify the tiltfiles and had to seperate the stacks into single-stack (ipv4) and dual-stack(ipv6) because I was unable to test ipv6 with Colima on M2 Pro.

I took the liberty of updating the Tiltfiles with latest chart versions.
- Removed metallb in favor of cilium with l2
- Started integration with cert-manager (i did not test to see if the tls routes and grpc routes actually respond only that they are added to k8s_gateway)
- Updating watched resources
- Used nerdctl instead of docker_restart
- i kept the makefile intact and left the docker step inside each tiltfiles.


I'll be committing all these changes

**Comment by networkop** [comment]
nice, thanks @larivierec , let me know when the PR is ready for review

**Comment by larivierec** [comment]
> nice, thanks @larivierec , let me know when the PR is ready for review

Oh, yeah you can go ahead.
I changed the status of this PR to `ready` I don't know if GH notififes you.

**Comment by networkop** [comment]
sorry @larivierec , been a bit busy. will try to review this week

File: examples/install-clusterwide.yml
Comment (networkop) [inline examples/install-clusterwide.yml:59]: why did you comment this out?

File: README.md
Comment (networkop) [inline README.md:290]: what's the purpose of `Tiltfile.single`?

Comment (networkop) [reply README.md:290]: why would't you just create a dedicated Tiltfile for lima VMs?

File: examples/install-clusterwide.yml
Comment (networkop) [inline examples/install-clusterwide.yml:73]: is this meant to be indented?

File: examples/install-clusterwide.yml
Comment (networkop) [inline examples/install-clusterwide.yml:83]: same question as above

File: examples/install-clusterwide.yml
Comment (networkop) [inline examples/install-clusterwide.yml:120]: same as above

File: gateway_test.go
Comment (networkop) [inline gateway_test.go:320]: why renaming this? is this going to be generic for all routes?

File: kubernetes.go
Comment (networkop) [inline kubernetes.go:72]: what would happen if you don't have GwAPI CRDs installed? is there a way to make sure we don't crash the pod when a cluster is missing gwAPI CRDs?

File: Tiltfile
Comment (networkop) [inline Tiltfile:39]: is there any reason not to use a stable Cilium version. L2 announcement should be in 1.14.X

**Review by networkop — CHANGES_REQUESTED** [review]
hey @larivierec , thanks for the massive PR and apologies for the delay in reviewing.
I've left a few comments, let me know what you think

File: examples/install-clusterwide.yml
Comment (larivierec) [reply examples/install-clusterwide.yml:59]: Mainly because by default I'm not sure people test gateway apis.

File: examples/install-clusterwide.yml
Comment (larivierec) [reply examples/install-clusterwide.yml:73]: All of these files were limited using vscode yaml linter

File: Tiltfile
Comment (larivierec) [reply Tiltfile:39]: yes actually, because 1.15.0-pre.1/2 support the latest gateway api crds where as the stable doesn't

File: examples/install-clusterwide.yml
Comment (larivierec) [reply examples/install-clusterwide.yml:120]: linter did this automatically

File: gateway.go
Comment (larivierec) [inline gateway.go:30]: same: linter

File: gateway_test.go
Comment (larivierec) [reply gateway_test.go:320]: i re-use the same function for all GRPC, TLS and HTTP routes
i could separate them if you'd like but fundamentally, they only need an IP Address.

File: kubernetes.go
Comment (larivierec) [reply kubernetes.go:72]: yeah, that's already handled by the existGatewayCRDs function

Here: [Function](https://github.com/ori-edge/k8s_gateway/blob/master/kubernetes.go#L189)

and [Handle Error](https://github.com/ori-edge/k8s_gateway/blob/master/kubernetes.go#L201)

currently, if gateway crds aren't installed, it'll just skip over the resources completely, which is what we want.

**Review by larivierec — COMMENTED** [review]
i reviewed your comments, see my replies if you can :) thanks

**Comment by larivierec** [comment]
i'll have another review for Gateway APIs v1.0.0 update after this one gets merged.
won't be too complicated to change.

File: README.md
Comment (networkop) [reply README.md:290]: perfect. so  there's one file for linux and one file for mac

File: Tiltfile
Comment (networkop) [reply Tiltfile:39]: ah, I see. that's fine then. just need to bump it to a stable once  1.15 is out

File: gateway_test.go
Comment (networkop) [reply gateway_test.go:320]: sounds reasonable. thanks

File: kubernetes.go
Comment (networkop) [reply kubernetes.go:72]: ah, I see.

File: examples/install-clusterwide.yml
Comment (networkop) [reply examples/install-clusterwide.yml]: ack. sounds good.

File: README.md
Comment (networkop) [reply README.md:290]: can you just update the readme and maybe rename the file to something reflecting that it's for mac or nerdctl.
otherwise single is a bit confusing.

**Comment by networkop** [comment]
@larivierec so apart from tiltfile updates this PR is good to be merged. 
Let's keep the GwAPI 1.0 updates to its own PR.

**Comment by larivierec** [comment]
Awesome, do you want me to rename the file to something else maybe?

Tiltfile.mac

**Comment by networkop** [comment]
yep. Tiltfile.mac sounds good.
and btw, since you've mentioned that Cilium now implements the GwAPI spec, would it make sense to drop Istio (from tiltfile)?

**Comment by larivierec** [comment]
> yep. Tiltfile.mac sounds good. and btw, since you've mentioned that Cilium now implements the GwAPI spec, would it make sense to drop Istio (from tiltfile)?

Ok for the tiltfile.
No I don't think so, from my perspective, it adds value as it shows the this plugin works with istio as well :)

I'll go ahead and make the changes and squash everything nicely.

File: README.md
Comment (larivierec) [reply README.md]: done, i opted for `Tiltfile.nerdctl` finally, it made more sense because nerdctl any os and the tiltfile isn't restricted to M powered macs

**Comment by larivierec** [comment]
everything should be good, let me know if I need to do something else.
thanks for the review 👍

**Comment by networkop** [comment]
awesome 🎉 
thanks for all the great work @larivierec

**Pull request merged**
